### ISSUE TITLE
[Feature Request] Automatic ClusterSpec Propagation for multiple hosts

### ISSUE THREAD

**Issue by UWFrank** [tensorflow/tensorflow#14232]
Currently, when launching a distributed TensorFlow job, user need to manually input all the worker hosts' IP and port number. This is not too convenient and does not scale well. It would be really nice to have the native TensorFlow functionality that workers can automatically register themselves on master service without knowing all the host IP and port beforehand. Not sure if there is existing solution to solve this problem. But I used some of the building blocks (ClusterSpec Propagation and ClusterResolver) to enable this feature on my client code. Please let me know if it is a good approach to do this, I'd love to contribute if there's interest in this functionality. 

The solution I have involve the following steps:
1) Master service start server with user specified port and wait for all workers to register
2) Worker register themselves by sending ClusterSpec to Master service and wait for response
3) Master service waits until numbers of workers registered matches requested worker number
4) Master service merges ClusterSpecs and propagates to all registered workers
5) Master service and workers continues

**Comment by mrry** [comment]
Passing on to @saeta for ClusterSpec propagation and @frankchn for ClusterResolver.

It would be neat if we could implement this without big changes to the runtime. For example, could we use a TF server to implement step 1, and short-lived sessions targeting the master from each of the workers to implement steps 2 and 3?

**Comment by frankchn** [comment]
Hi @UWFrank, is your code available as open source anywhere? We are happy to take a look at it.

**Comment by UWFrank** [comment]
Hi @frankchn, my design would require a new API for creating the server. Parameters such as num_ps_hosts, num_worker_hosts, job_name and task_index need to be provided by the user. I am working on a patch right now and can send a PR once done.

**Comment by frankchn** [comment]
Hi @UWFrank, would it be possible to write up a short Google Doc/Markdown doc about any proposed API changes? We try to review API changes early in the design process so to avoid extensive rewrites, etc...

**Comment by yongtang** [comment]
@UWFrank This is an area of interest to us as well. Please keep us posted and let me know if I can be of any help in this area.

**Comment by UWFrank** [comment]
Hello @frankchn I've drafted a [design proposal](https://docs.google.com/document/d/18XgUwmmxvhR-sk9A2KFD8Wom8H0sfI7PKtacVI_7jvc) that include major API changes, please let me know your thoughts. @yongtang thank you for your interest, please also feel free to comment on the design doc.

**Comment by yaroslavvb** [comment]
@UWFrank In a async SGD cluster there's typically two jobs, `job_name='ps'` and `job_name='worker'`. Your API suggests to use following to create coordinator:
`create_distributed_server(num_workers, job_name, task_index, master_host, is_master=False)`

Does this mean you would create two distributed servers, one for each job?

For efficient async SGD, the number of live hosts in worker job should be allowed to change over time without affecting session creation, or session run calls of remaining live hosts. This allows training to proceed even if some hosts are unavailable. Currently this is achieved using sparse job config (https://github.com/tensorflow/tensorflow/commit/8177edd7700ccbe0c831e680a1acb5275819d762). The cluster propagation design seems to preclude this use-case.

For robust async training, example here's what some ClusterSpecs looks like for some servers with 4 parameter server shards and 2 workers:

```
job_name=worker, task_id=1:
{'worker': {1: '172.31.35.121:3000'}, 'ps': ['172.31.42.219:3000', '172.31.38.22:3000', '172.31.32.142:3000', '172.31.35.253:3000']})

job_name=worker, task_id=2
{'worker': {2: '172.31.41.31:3000'}, 'ps': ['172.31.42.219:3000', '172.31.38.22:3000', '172.31.32.142:3000', '172.31.35.253:3000']})

job_name=ps, task_id=3:
{'worker': ['172.31.40.171:3000', '172.31.35.121:3000', '172.31.41.31:3000', '172.31.36.117:3000', '172.31.38.4:3000'], 'ps': {3: '172.31.35.253:3000'}})
```

Each `tf.train.Server` instance (worker or ps) knows about itself. Worker servers need to know locations of all ps shards, but not of other workers. PS servers need to know about all workers but not other ps shards. As long as ps servers are only used as [workers](https://github.com/tensorflow/tensorflow/blob/9089ab59826ce656b6f8026b08471ca6696cacf9/tensorflow/core/protobuf/worker.proto), and not as [masters](https://github.com/tensorflow/tensorflow/blob/2eac53d5ea540b0b09326ba9330b6051d742532d/tensorflow/core/protobuf/master.proto), the number of live worker servers can change without affecting training. On other hand, training will stop if any ps server goes down.

**Comment by UWFrank** [comment]
@yaroslavvb I've updated the draft to address your comment to support sparse job config.
I added a new parameter 'sparse_jobs_config' to allow users to specify related job's task_index so that only those host addresses get propagated back from the master server.

> Does this mean you would create two distributed servers, one for each job?

Each **task** launch their own server.There should only be one master for each cluster. I should clarify that the worker referenced in 'num_workers' in the parameter is not job_name='worker', but type of server.

Please let me know your thoughts.

Thanks

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by yaroslavvb** [comment]
relevant: https://github.com/tensorflow/tensorflow/issues/14232

**Comment by hungj** [comment]
Quick update, @UWFrank's implementation is almost code complete, we'll be doing some minor tweaks / testing and will target posting the implementation sometime early next month.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @frankchn, @saeta: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 44 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by mistersdevelopment** [comment]
Has any progress been made regarding this issue or #11896?

**Comment by redfish-lee** [comment]
@hungj @UWFrank is there any updates?

**Comment by tensorflowbutler** [comment]
It has been 29 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 44 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by reeze** [comment]
This is an important feature when building scalable ml systems.  is it still active?

**Comment by redfish-lee** [comment]
@UWFrank  What's `ManagedTrainingSession` mentioned in [README](https://github.com/tensorflow/tensorflow/blob/r1.8/tensorflow/contrib/cluster_resolver/README.md) exactly refers to ?
Is it a session which we can pass clusterSpec in session creation?

Thanks

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 30 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
We are closing this issue for now due to lack of activity. Please comment if this is still an issue for you. Thanks!

**Comment by gaocegege** [comment]
Hi is there any update? I think it is an important feature for the community.

**Comment by frankchn** [comment]
The new way to use TensorFlow for distributed training is called distribution strategies: https://github.com/tensorflow/tensorflow/blob/r1.12/tensorflow/contrib/distribute/README.md, so we do encourage you to take a look at those examples.

**Comment by liyi193328** [comment]
Is there any updates?

**Comment by fesun** [comment]
In case someone else is also confused by this, TF has the ability to propagates cluster spec automatically. For example the ps servers can boot without knowing any worker devices.
See #38519
